src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -338,9 +338,9 @@ private Runnable getPostRequestRunnable(int numberOfEvents) {
                     LOGGER.error("BatchEmitter failed to send {} events: code: {}", eventsInRequest.size(), code);
                     eventStore.cleanupAfterSendingAttempt(true, batchedEvents.getBatchId());
 
-                    // exponentially increase retry backoff time after the first failure
-                    if (!retryDelay.compareAndSet(0, 50L)) {
-                        retryDelay.updateAndGet(currentDelay -> currentDelay * 2);
+                    // exponentially increase retry backoff time after the first failure, up to the maximum wait time
+                    if (!retryDelay.compareAndSet(0, 100)) {
+                        retryDelay.getAndSet(calculateRetryDelay());
```

Comment:
I'm not totally sure this block is atomic. The `calculateRetryDelay` is not executed in a synchronized block so we can't assure that retryDelay doesn't change during the computation. The `getAndSet` just set the value and return the previous one but we actually don't use the returned value so it's technically like `retryDelay.set(calculateRetryDelay());`.

Instead, `updateAndGet` executes the block in the argument passing the current value. It checks that the value doesn't change meanwhile ([implementation](https://github.com/JetBrains/jdk8u_jdk/blob/94318f9185757cc33d2b8d527d36be26ac6b7582/src/share/classes/java/util/concurrent/atomic/AtomicInteger.java#L237)).
Even if not synchronized it allows the atomicity.

Maybe something like `retryDelay.updateAndGet(currentDelay -> calculateRetryDelay(currentDelay));` should work, or if possible `retryDelay.updateAndGet(calculateRetryDelay);`.
In this case `calculateRetryDelay` should take `currentDelay` as an argument.

Reply:
Aha thanks, thought it didn't look quite right atomically! I'll investigate `updateAndGet`